### PR TITLE
EVG-20278 Reorganize setup.initial for task setup

### PR DESCRIPTION
**Pull request by malikchaya2**
EVG-20278

### Description
This moves a bunch of logic that is conceptually part of setting up the task into a setupTask function. It also removes some things that are not needed because they are duplicates. 

### Testing
Added a test and fixed existing tests

### CODE REVIEW

File: globals.go
Comment (Kimchelly) [inline globals.go:323]: Remove "default" since it's not a configurable value - the app server will system-fail a task that exceeds this.

File: agent/agent.go
Comment (Kimchelly) [inline agent/agent.go:629]: `// Setup occurs before the task is actually running, so it's not abortable. If setup is taking a long time, timing it out after the heartbeat timeout ensures the operation will give up within the same timeout as the usual amount of time a task is allowed to run before it must heartbeat.`

File: agent/agent.go
Comment (Kimchelly) [inline agent/agent.go:471]: Seems like this can just return true rather than setting a variable to true and returning it.

File: agent/agent.go
Comment (Kimchelly) [inline agent/agent.go:481]: Same thing as above.

File: agent/agent.go
Comment (Kimchelly) [reply agent/agent.go:471]: We should also explicitly return false for shouldExit, rather than assume that another piece of code higher up will set it for us. Generally, returning the named output parameter is a code smell because it makes it difficult to deduce when another unrelated return statement changes it value due to setting it.

File: agent/agent.go
Comment (Kimchelly) [inline agent/agent.go:493]: Same thing as above.

File: agent/agent.go
Comment (Kimchelly) [inline agent/agent.go:508]: Same thing as above.

File: agent/agent.go
Comment (Kimchelly) [inline agent/agent.go:518]: It seems like the value of `setupFailed` is always equivalent to the returned error being non-nil. Can we consolidate the boolean?

File: agent/agent.go
Comment (Kimchelly) [inline agent/agent.go:434]: I would put a comment above this stating what this is generally trying to accomplish (like initialize the logger and load the task config data).

Comment (Kimchelly) [inline agent/agent.go:434]: nit: rather than just call it something like `tcInput`, something like `initialTC` or whatnot might be more illustrative that it's not just an input but also is the "seed" task context.

File: agent/agent.go
Comment (Kimchelly) [inline agent/agent.go:470]: This return logic (and the others that are the same in this function) seem like they're copy-pasting the same logic. Can we consolidate them into a helper function?

File: agent/agent.go
Comment (Kimchelly) [inline agent/agent.go:713]: Since we aren't absorbing some things that apply only to pre and main (like the stats collector) in `runPreAndMain`, can we make a small helper function to also wrap the setup logic in `runPreAndMain`?

File: agent/agent.go
Comment (Kimchelly) [inline agent/agent.go:451]: It seems like this `tc.jasper` has no usages. Can we delete the field entirely?

File: agent/agent.go
Comment (Kimchelly) [inline agent/agent.go:450]: It seems like this is trying to clear out any of the old processes from the previous task. Rather than doing that here (when we're setting up a new task), can we do it in `finishPrevTask` (because we're already tying up loose ends from the previous task)?

File: agent/agent_test.go
Comment (Kimchelly) [inline agent/agent_test.go:781]: Remove comment to self, or if it's intended to stay, make this comment a little more meaningful.

File: agent/agent_test.go
Comment (Kimchelly) [inline agent/agent_test.go:805]: nit: remove random newline.

File: agent/agent_test.go
Comment (Kimchelly) [inline agent/agent_test.go:806]: Do we have to test the proper handling of the error conditions in `setupTask` (such as getting bad data about a task, or the context erroring while it's setting up)? Might need to tweak the mock communicator to return fake errors.

File: agent/agent_test.go
Comment (Kimchelly) [inline agent/agent_test.go:869]: "in the same version and task group name"?

**Review by Kimchelly — CHANGES_REQUESTED** [review]
Did you test this in staging and verify that the logs at the beginning of the task look appropriate?

File: agent/agent.go
Comment (malikchaya2) [reply agent/agent.go:471]: Returned true. But shouldExit should be whatever handleTaskResponse returns.

File: agent/agent.go
Comment (malikchaya2) [reply agent/agent.go:518]: a.handleTaskResponse overwrites err, so we can't rely on err being non-nil

File: agent/agent.go
Comment (malikchaya2) [reply agent/agent.go:713]: Is there enough that we can move over that would make it worthwhile? It seems like it would just be the stats collector.

File: agent/agent_test.go
Comment (malikchaya2) [reply agent/agent_test.go:806]: I don't know how crucial that is because it's not doing too complex error handling and it's mostly moved from other functions that didn't test it either. The mock communicator is not set up for this to be tested easily. If you feel strongly about it I can try to add it.

File: agent/agent.go
Comment (Kimchelly) [reply agent/agent.go]: I still have this question.

File: agent/agent.go
Comment (Kimchelly) [inline agent/agent.go:503]: `setupCtx` should be the first parameter for Go conventions.

I would also order it as `ctx, tc, err, shouldUseFallbackLogger` since we generally pass in `tc` to a lot of functions.

Comment (Kimchelly) [inline agent/agent.go:503]: nit: I would call `shouldSetLogger` something more meaningful like `shouldUseFallbackLogger` (i.e. the logger when we can't set up the logger). Alternatively, initializing and passing in the logger in the error cases where the logger is not set up might be cleaner than a bool.

File: agent/agent.go
Comment (Kimchelly) [reply agent/agent.go:706]: Hmm, I guess it's not that much setup logic now that stuff has moved out. Never mind!

File: agent/agent_test.go
Comment (Kimchelly) [reply agent/agent_test.go:791]: Eh, we'd ideally test it since we want to know if end in tasks system failure as expected to indicate that there's a bug (rather than the agent doing something undesirable like run the task forever, silently succeed, panic the whole agent, or something else).

If it's way too difficult though, it's alright to skip.

File: agent/agent.go
Comment (Kimchelly) [inline agent/agent.go:503]: I would name the output parameters since they're not obvious.

File: agent/agent.go
Comment (Kimchelly) [reply agent/agent.go:500]: It seems like `setupTask` should return an error though (even if `handleTaskResponse` doesn't produce an error). We can simplify and remove the `setupFailed` boolean by using an error catcher inside `handleSetupError`, catching the error passed into `handleSetupError`, and then catching the error from `handleTaskResponse`, so that returning an error from `setupTask` is equivalent to `setupFailed`.

File: agent/agent.go
Comment (Kimchelly) [inline agent/agent.go]: This can also call `a.handleSetupError`.

File: agent/agent.go
Comment (Kimchelly) [inline agent/agent.go:475]: I don't think we actually need this log because the error is already passed to `handleSetupError`, which could just log the error to the `tc.logger` in there (this would work for the earlier cases as well, where the logger is not set up and we use the fallback one).

**Review by Kimchelly — CHANGES_REQUESTED** [review]

**Comment by Kimchelly** [comment]
I still also have the staging question.

File: agent/agent.go
Comment (malikchaya2) [reply agent/agent.go]: I removed it, it's not there anymore

File: agent/agent_test.go
Comment (malikchaya2) [reply agent/agent_test.go:791]: yeah I think it would be difficult. We can maybe create a ticket to do it separately.

File: agent/agent.go
Comment (malikchaya2) [reply agent/agent.go:475]: but then we'd start logging errors that we weren't logging before like "making task config" and "setting up logger producer"

File: agent/agent.go
Comment (malikchaya2) [reply agent/agent.go]: The complexity it would add to not 
`grip.Infof("Task complete: '%s'.", tc.task.ID)`
and instead `tc.logger.Execution().Infof("Stopping task execution: %s", err)` made it feel not worth it.

File: agent/agent.go
Comment (malikchaya2) [reply agent/agent.go:503]: it's now only one bool that's named when it's returned

Comment (malikchaya2) [reply agent/agent.go:503]: I moved `tc.logger = client.NewSingleChannelLogHarness("agent.error", a.defaultLogger)` out so that bool is not needed.

File: agent/agent.go
Comment (malikchaya2) [inline agent/agent.go:485]: I tested it on staging. I had to add this back in because the otherwise it didn't log like it does now. It's because the first one is before startLogging. I can probably remove the first one but I'm too scared it will have some hidden unintended consequence. 

The rest of the logs are all there. The order differs each time tasks are run so it's hard to compare the order to main.

Comment (malikchaya2) [reply agent/agent.go:485]: the logs on main: https://parsley.mongodb.com/evergreen/evergreen_build_simple_test_patch_1072528ebb0ab31bda6733e16a6ff8bbd19651e7_64df9cada4cf4733f348ed92_23_08_18_16_30_45/1/all?bookmarks=0,1,2,3,4,5,6,7,417

the logs with the changes: 
https://parsley-staging.corp.mongodb.com/evergreen/evg_build_simple_test_patch_1072528ebb0ab31bda6733e16a6ff8bbd19651e7_64dfaa4f9dbe323bd245bf8c_23_08_18_17_28_59/0/all?bookmarks=0,1,2,3,4,5,6,7,213

Comment (malikchaya2) [reply agent/agent.go:485]: while the order is different between main and with the changes, they are also not in the same order with every run. 
for example, here is another run with the changes, and the order is different than the other staging link above: 

https://parsley-staging.corp.mongodb.com/evergreen/evg_build_simple_test_patch_1072528ebb0ab31bda6733e16a6ff8bbd19651e7_64dfafc19dbe323bd245bf96_23_08_18_17_52_12/0/all?bookmarks=0,210

Comment (Kimchelly) [reply agent/agent.go:485]: SGTM. Don't really know why the logs would be out of order, but I have noticed before that logs that come in at the same time with the same exact timestamp oftentimes are mixed up.

File: agent/agent.go
Comment (Kimchelly) [inline agent/agent.go:520]: I would `catcher.Wrap(err, "handling task response")` (same with the error caught above) so that we can distinguish between this error and the error caught above this. Otherwise, it'll be weird because the resolved catcher errors, when logged, will look like one error happened (the setup error), then it caused a completely random set of errors happened (but it didn't because it's a separate error).

File: agent/agent_test.go
Comment (Kimchelly) [reply agent/agent_test.go:790]: I'll leave it to you to make the ticket.

File: agent/agent.go
Comment (Kimchelly) [inline agent/agent.go:458]: Can we actually `grip.Alert` in this case, and potentially continue on error? This error case should never happen unless there's a programmatic bug in the agent command registration setup. It's also the case that `setup.initial` is just used to populate the "Failing command" during setup, where it's not actually running any actual agent commands.

File: agent/agent.go
Comment (Kimchelly) [inline agent/agent.go:463]: nit: there's no need to declare this variable because you already define it in L464.

File: agent/agent.go
Comment (Kimchelly) [inline agent/agent.go:488]: Like above, I'm pretty sure we can continue on error if `setup.initial` is not registered.

File: agent/agent_test.go
Comment (Kimchelly) [inline agent/agent_test.go:740]: mega nit: since you indent with one space in the other blocks, should also indent here (although YAML generally has 2 or 4 spaces for indentation, not 1).

**Review by Kimchelly — CHANGES_REQUESTED** [review]

File: agent/agent_test.go
Comment (malikchaya2) [reply agent/agent_test.go:790]: https://jira.mongodb.org/browse/EVG-20708

File: agent/agent.go
Comment (malikchaya2) [reply agent/agent.go]: :see_no_evil:  true

File: agent/agent.go
Comment (Kimchelly) [reply agent/agent.go]: Seems like this can use `handleSetupError` now that it logs both.

File: agent/agent.go
Comment (Kimchelly) [reply agent/agent.go]: If this is going to alert, we have to do `else { tc.setCurrentCommand(factory()) }` since the factory will be nil when `!ok`.

File: agent/agent.go
Comment (Kimchelly) [reply agent/agent.go]: Same thing as above about checking `if factory != nil` before dereferencing.

**Review by Kimchelly — APPROVED** [review]
I left a few small comments, but you can easily resolve them without another review.

**Pull request merged**
